sdks/python/apache_beam/metrics/cells.py

```diff
@@ -0,0 +1,200 @@
+        self._state = DirtyState.CLEAN
+
+  def before_commit(self):
+    with self._lock:
+      if self._state == DirtyState.CLEAN:
+        return False
+      else:
+        self._state = DirtyState.COMMITTING
+        return True
+
+
+class MetricCell(object):
+  """ Base class of a Cell that tracks the state of a metric.
+
+  It's thread safe.
+  """
+  def __init__(self):
+    self.dirty = DirtyState()
+    self._lock = threading.Lock()
+
+  def get_cumulative(self):
+    raise NotImplementedError
+
+
+class CounterCell(Counter, MetricCell):
+  """ Tracks the state of a counter metric.
+
+  It's thread safe.
+  """
+  def __init__(self, *args):
+    super(CounterCell, self).__init__(*args)
+    self.value = 0
+
+  def combine(self, other):
+    result = CounterCell()
+    result.inc(self.value + other.value)
+    return result
+
+  def inc(self, n=1):
+    with self._lock:
+      self.value += n
+      self.dirty.modified()
+
+  def get_cumulative(self):
+    return self.value
+
+
+class DistributionCell(Distribution, MetricCell):
+  """ Tracks the state of a distribution metric.
+
+  It's thread safe.
+  """
+  def __init__(self, *args):
+    super(DistributionCell, self).__init__(*args)
+    self.data = DistributionData(0, 0, None, None)
+
+  def combine(self, other):
+    result = DistributionCell()
+    result.data = self.data.combine(other.data)
+    return result
+
+  def update(self, value):
+    with self._lock:
+      self.dirty.modified()
+      self._update(value)
+
+  def _update(self, value):
+    self.data.count += 1
+    self.data.sum += value
+    self.data.min = (value
+                     if self.data.min is None or self.data.min > value
+                     else self.data.min)
+    self.data.max = (value
+                     if self.data.max is None or self.data.max < value
+                     else self.data.max)
+
+  def get_cumulative(self):
+    return self.data
+
+
+class DistributionData(object):
+  def __init__(self, sum, count, min, max):
+    self.sum = sum
+    self.count = count
+    self.min = min
+    self.max = max
+
+  def __eq__(self, other):
+    return (self.sum == other.sum and
+            self.count == other.count and
+            self.min == other.min and
+            self.max == other.max)
+
+  def __neq__(self, other):
+    return not self.__eq__(other)
+
+  def __repr__(self):
+    return '<DistributionData({}, {}, {}, {})>'.format(self.sum,
+                                                       self.count,
+                                                       self.min,
+                                                       self.max)
+
+  @property
+  def mean(self):
+    return self.sum/self.count
+
+  def combine(self, other):
+    if other is None:
+      return self
+    else:
+      return DistributionData(
+          self.sum + other.sum,
+          self.count + other.count,
+          min(x for x in (self.min, other.min) if x is not None),
```

Comment:
What if self.min and other.min are both None?  This code will fail with "ValueError: min() arg is an empty sequence".